Sources/Plasma/CoreLib/hsSystemInfo.cpp

```diff
@@ -0,0 +1,355 @@
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "hsSystemInfo.h"
+
+#include "hsConfig.h"
+#include "plFileSystem.h"
+#include "hsStream.h"
+#include "hsWindows.h"
+
+#include <cstring>
+#include <iterator>
+#include <optional>
+#include <string_theory/string>
+#include <stdio.h>
+
+#ifdef HAVE_SYSINFO
+#    include <sys/sysinfo.h>
+#    include <sys/utsname.h>
+#endif
+
+#ifdef HAVE_SYSCTL
+#    include <sys/types.h>
+#    include <sys/sysctl.h>
+#endif
+
+#ifdef HS_BUILD_FOR_UNIX
+#    include <sys/utsname.h>
+#endif
+
+#ifdef HS_BUILD_FOR_APPLE
+#    include <CoreFoundation/CoreFoundation.h>
+
+    extern "C" {
+        CF_EXPORT const CFStringRef _kCFSystemVersionProductNameKey;
+        CF_EXPORT const CFStringRef _kCFSystemVersionProductVersionKey;
+
+        CF_EXPORT CFDictionaryRef _CFCopyServerVersionDictionary();
+        CF_EXPORT CFDictionaryRef _CFCopySystemVersionDictionary();
+    }
+#endif
+
+#ifdef HAVE_CPUID
+#    if defined(_MSC_VER) || ((defined(_WIN32) || defined(_WIN64)) && defined(__INTEL_COMPILER))
+#        include <intrin.h>
+#        define MSC_COMPATIBLE
+#        define cpuid_t int
+#    elif defined(__GNUC__)
+#        include <cpuid.h>
+#        define GCC_COMPATIBLE
+#        define cpuid_t unsigned int
+#    else
+#        define cpuid_t int
+#    endif
+#endif
+
+ST::string hsSystemInfo::AsString()
+{
+    ST::string_stream ss;
+    ss << "OS: " << GetOperatingSystem() << '\n';
+    ss << "CPU: " << GetCPUBrand() << '\n';
+    ss << "RAM: " << GetRAM() << " MiB";
+    return ss.to_string();
+}
+
+static inline void ICPUID(cpuid_t* info, int function_id)
+{
+#if defined(MSC_COMPATIBLE)
+    __cpuid(info, function_id);
+#elif defined(GCC_COMPATIBLE)
+    __get_cpuid(function_id, &info[0], &info[1], &info[2], &info[3]);
+#elif !defined(HAVE_CPUID)
+    memset(info, 0, sizeof(info));
+#endif
+}
+
+ST::string hsSystemInfo::GetCPUBrand()
+{
+    cpuid_t cpuInfo[4]{};
+    ICPUID(cpuInfo, 0x80000000);
+    unsigned int nExIds = (unsigned int)cpuInfo[0];
+    const unsigned int brandIds[]{ 0x80000002, 0x80000003, 0x80000004 };
+
+    char str[std::size(brandIds) * sizeof(cpuInfo)]{};
+    char* strp = str;
+    for (unsigned int i : brandIds) {
+        if (i <= nExIds) {
+            ICPUID(reinterpret_cast<cpuid_t*>(strp), i);
+            strp += sizeof(cpuInfo);
+        }
+    }
+
+    ST::string result(str);
+    if (result.empty())
+        result = ST_LITERAL("Unknown");
+    return result;
+}
+
+#ifdef HS_BUILD_FOR_APPLE
+static inline bool IGetAppleVersion(ST::string& system)
+{
+    CFDictionaryRef dict = _CFCopyServerVersionDictionary();
+    if (dict == nullptr)
+        dict = _CFCopySystemVersionDictionary();
```

Comment:
Any particular reason for using these undocumented functions rather than reading the dictionaries from the plists (ServerVersion.plist and SystemVersion.plist in /System/Library/CoreServices)? The plists are at least somewhat documented and widely known, so they are unlikely to go away, whereas I couldn't find any information about these internal functions outside of the CoreFoundation source code. (The undocumented functions actually just read the plists internally, so it shouldn't make a difference.)

Helpful example code for how to do the plist reading: https://stackoverflow.com/a/56403143

Reply:
Per @dpogue, Apple does shenanigans if you try to open that plist file: https://stackoverflow.com/questions/69097567/macos-version-returned-as-10-16-instead-of-12-0 - we've found that many other projects use these undocumented APIs to do the same job in a reliable, reproducible way. Hence this code.

Reply:
Ugh, compatibility hacks... According to [this JDK bug](https://bugs.openjdk.java.net/browse/JDK-8269850), you can read .SystemVersionPlatform.plist to get the real version number on macOS 11 and later, and if that doesn't exist, fall back to the regular SystemVersion.plist.

And if you're calling `sw_vers`, you apparently have to pass the environment variable `SYSTEM_VERSION_COMPAT=0`, or it will also return 10.16 instead of 11.0.

Reply:
What Hoikas is doing here is actually the same as what `sw_vers` does (at least, according to the open-source sw_vers code that Apple published): https://opensource.apple.com/source/DarwinTools/DarwinTools-1/sw_vers.c.auto.html

Reply:
Yep, Apple likes to use undocumented APIs in their applications without providing a documented equivalent for external developers 🙂

Yeah, these functions are not *that* likely to go away, especially if other major projects use them as well. I just think that in a crash handler, it's better to be extra safe and avoid undocumented APIs, just in case they do change and end up crashing the crash handler somehow.